sql-plugin/src/main/java/com/nvidia/spark/rapids/JCudfUtil.java

```diff
@@ -0,0 +1,437 @@
+    return !cudfRowOffsetMock.hasVarSizeData() &&
+        cudfRowOffsetMock.getCurrentByteOffset() < JCUDF_MAX_FIXED_ROW_SIZE_FITS_OPTIMIZED;
+  }
+
+  /**
+   * A helper class to calculate the columns, validity, and variable width offsets.
+   * This helper is used to get an estimate size for the row including variable-sized rows.
+   * The implementation does not make any assumptions whether the schema is packed or not.
+   * Unlike {@link JCudfRowVisitor}, this implementation does not cache any calculations.
+   */
+  public static class JCudfRowOffsetsEstimator {
+    Attribute[] attributes;
+    int byteCursor = 0;
+    int varSizeColIndex;
+    int validityBytesOffset = 0;
+
+    private JCudfRowOffsetsEstimator(Attribute[] attrArr) {
+      this.attributes = attrArr;
+      resetMeta();
+    }
+
+    private int calcColOffset(int ind) {
+      int res;
+      Attribute attr = attributes[ind];
+      DataType dataType = attr.dataType();
+      DType rapidsType = GpuColumnVector.getNonNestedRapidsType(dataType);
+      if (isFixedLength(dataType)) {
+        int length = rapidsType.getSizeInBytes();
+        res = alignOffset(byteCursor, length);
+        byteCursor = res + length;
+        return res;
+      }
+      if (!hasVarSizeData()) {
+        // set the index to the first variable size column.
+        varSizeColIndex = ind;
+      }
+      res = alignOffset(byteCursor, JCUDF_VAR_LENGTH_FIELD_ALIGNMENT);
+      byteCursor = res + JCUDF_VAR_LENGTH_FIELD_SIZE;
+      return res;
+    }
+
+    /**
+     * Sets the column offsets
+     * @param offsets array containing the offset for all columns
+     * @return the current byte offset of the row.
+     */
+    public int setColumnOffsets(int[] offsets) {
+      resetMeta();
+      for (int i = 0; i < attributes.length; i++) {
+        offsets[i] = calcColOffset(i);
+      }
+      return byteCursor;
+    }
+
+    private void resetMeta() {
+      varSizeColIndex = attributes.length;
+      byteCursor = 0;
+    }
+
+    private void setValidityBits(int fieldsCount){
+      validityBytesOffset = byteCursor;
+      byteCursor += calculateBitSetWidthInBytes(fieldsCount);
+    }
+
+    public int getEstimateSize() {
+      resetMeta();
+      for (int i = 0; i < attributes.length; i++) {
+        calcColOffset(i);
+      }
+      // set the validity size
+      setValidityBits(attributes.length);
+      if (hasVarSizeData()) {
+        for (int ind = varSizeColIndex; ind < attributes.length; ind++) {
+          DataType dataType = attributes[ind].dataType();
+          if (isFixedLength(dataType)) {
+            continue;
+          }
+          DType rapidsType = GpuColumnVector.getNonNestedRapidsType(dataType);
+          addVarSizeData(rapidsType);
+        }
+      }
+      byteCursor = alignOffset(byteCursor, JCUDF_ROW_ALIGNMENT);
+      return byteCursor;
+    }
+    private int addVarSizeData(DType rapidsType) {
+      int length = getEstimateSizeForVarLengthTypes(rapidsType);
```

Comment:
What happens if this estimate is low? Is this used for buffer allocation?

Reply:
This handled as described in https://github.com/NVIDIA/spark-rapids/pull/5998#issuecomment-1191875220